src/python/gem5/simulate/simulator.py

```diff
@@ -463,6 +463,23 @@ def get_id(self) -> Optional[str]:
 
         return None
 
+    def set_on_exit_handler(
+        on_exit_event: Dict[
+            ExitEvent,
+            Union[
+                Generator[Optional[bool], None, None],
+                List[Callable],
+                Callable,
+            ],
+        ]
+    ) -> None:
+        """Set the on_exit_event handler for the simulator.
+
+        Set the exit event handler that specifies what to execute on each exit
+        event."""
+
+        self._on_exit_event = on_exit_event
+
```

Comment:
You'll need to do some translation of the dictionary like we do in the constructor (here: https://github.com/gem5/gem5/blob/186a913a48f13bdd484fcdef17ac3e28d2b8b4c9/src/python/gem5/simulate/simulator.py#L363).

I'd recommend having the constructor call this new function and moving the translation code here.